Experiment, parallelize some tests

majocha opened the pull request:
To run tests in parallel we must deal with global resources and global state accessed by the test cases.

**Out of proc:**
Tests running as separate processes are sharing the file system. We must make sure they execute in their own temporary directories and don't overwrite any hardcoded paths. This is already done, mostly in separate PR.

**Hosted:**
Many tests use hosted compiler and `FsiEvaluationSession`, sharing global resources and global state within the runner process:

- [x]  Console streams - this is swept under a rug for now by using a simple AsyncLocal stream splitter.
- [x] `FileSystem` global mutable of the file system shim - few tests that mutate it, must be excluded from parallelization.
- [x] `Environment.CurrentDirectory` - many tests executing in hosted session were doing a variation of `File.WriteAllText("test.ok", "ok")` all in the current directory i.e. `bin`, leading to conflicts. This is replaced with a threadsafe mechanism.
- [x] Environment variables, Path - mostly this applies to `DependencyManager`, excluded from parallelization for now.
- [x] `Async` default cancellation token - few tests doing `Async.CancelDefaultToken()` must be excluded from parallelization.
- [x] global state used in conjunction with `--times` option - tests excluded from parallelization.
- [ ] global mutable state in the form of multiple caches implemented as `ConcurrentDictionary`. This needs further investigation.

I'll ad to the above list if I recall anything else.



**Problems:**
Tests depending on tight timing, orchestrating stuff by combinations of `Thread.Sleep`, `Async.Sleep` and wait timeouts.
These are mostly excluded from parallelization, some attempts at fixing things were made.


**Obscure compiler bugs** revealed in this PR:

- `Internal error: value cannot be null` this mostly happens in coreClr, one time, sometimes a few times during the test run. 

- `Error creating evaluation session` because of NRE somewhere in `TcImports.BuildNonFrameworkTcImports`. This is more rare but may be related to the above.

These were related to some concurrency issues; modyfing `frameworkTcImportsCache` without lock and a bug in custom lazy implementation in il.fs. Hopefully both fixed now.

**Running in parallel:**
Xunit runners are configured with mostly default parallelization settings.

`dotnet test .\FSharp.sln -c Release -f net9.0` will run all discovered test assemblies in parallel as soon as they're built.
This can be limited with the `-m` switch. For example,
`dotnet test -m:2 .\FSharp.Compiler.Service.sln`
will limit the test run to at most 2 simultaneous processes. Still, each test host process runs its test collections in parallel. 

Some test collections are excluded form parallelization with `[<Collection(nameof DoNotRunInParallel)>]` attribute.

Running in the IDE with "Run tests in parallel" enabled will respect `xunit.runner.json` settings and the above exclusions.


TODO:
- [x]  Make sure this keeps working properly with `BUILDING_USING_DOTNET` scenario (#14677)

majocha commented:
Yeah this will not do much.

A lot of test cases write to stdout. We do `captureConsoleOutputs` in `CompilerAssert` but in a way that does not allow parallel execution. Not to mention in a lot of places there are `printfn` calls that will mangle outputs even more. All of those tests must run sequentially.

This needs a systemic approach to deal with stdout capture, maybe xUnit have some mechanism to do this in parallel.

psfinaki commented:
@majocha guess what, I was going to try the same today :) 

Thanks for looking into that. Indeed, it seems like we should just damn stop printing everything to the console, it's likely an artifact of older testing approaches. I cannot see any reason to do this instead of just in memory processing.

majocha commented:
xUnit does not run tests from the same module in parallel. It also does not parallelize `Theories`.
This can slow things down even with parallel execution enabled. In FSharpSuite we have modules with lots of slow tests.

This can be mitigated by customizing xUnit in code, I think?

psfinaki commented:
By customizing xUnit you mean setting up some special runner settings and assembly attributes?

We can do that. However - I am not a fan of this idea. xUnit's philosophy is really to apply good coding practices to tests. As in, write tests as you write code. Hence, e.g. compared to Nunit, it offers a very limited test platform voodoo (think fixtures, setup/teardown and so on), instead making as much as possible of the builtin language capabilities.

And so I would instead prefer keeping up with this philosophy. If, by default, xUnit parallelizes execution on the module level, then we should actually split modules into smaller ones - thereby it will improve code clarity and will generally add to better code organization :)

What do you think?

majocha commented:
By customizing I meant something like https://www.meziantou.net/parallelize-test-cases-execution-in-xunit.htm

But this is not the most pressing thing and probably not needed if splitting modules would do.


The biggest hurdle for now is correctly isolating the console when running tests in parallel. Redirecting with `Console.SetOut` for each test won't work anymore when another thread can also redirect it unpredictably.

Console writes come from multiple sources:

1. In process executing fsi, probably also fsc
2. Various `printfn` and `log` calls sprinkled through the test cases and helper code.
3. Source code compiled and executed in `AssemblyLoadContext` (or `AppDomain` in case of net472).

While we can manage 1. and 2., 3. is a bit harder.

majocha commented:
I've been chatting with Bing / Copilot about it, and it actually proposed a not bad idea:

Don't redirect the Console at all for individual tests. Instead install a custom thread splitting `TextWriter` upfront.
That splitting writer will keep a `ThreadLocal` inner `TextWriter` and direct all writes to it.

psfinaki commented:
Thanks for the further investigations here.

In the spirit of my comment above - I just vote for reinventing as few wheels as possible and removing those we've already reinvented here :) Unit tests rarely need any output at all, but if they do - it's good to use those [few means](https://xunit.net/docs/capturing-output) that xUnit provides for this, which are basically "plug in the writer if and when you need to".

I think it aligns with your thoughts above? It's important to make gradual changes here, probably actually in the way you outline it above. The current direction you're taking (removing stuff) looks promising!

---

Note, I am off until Monday with limited internet connection so cannot play with the code myself. Also, we've discussed this PR internally yesterday and were all very happy that things are moving in this space!

majocha commented:
This is at a state that can be run locally in VS test explorer or from the console with `build -c Release -testCoreClr`
I think it's shaving a few minutes from build -testCoreClr locally, this could be further improved by breaking up large modules in ComponentTests and FSharpSuite.

In the CI there's that weird `TaskCancelledException` in random tests, no idea where it comes from.

majocha commented:
I see some improvements.
For example in FSharp.sln full parallel test run of all net9.0 test projects takes 6.6 min on my PC. After breaking some large modules into smaller chunks, rather randomly, it's down to 4.3 min:
![image](https://github.com/user-attachments/assets/e99d38c2-3212-4df1-8154-1082bc11f9b2)

majocha commented:
Apart from isolating the console, we would need to isolate the other shared resource - filesystem.
Weeding out test cases writing to non-unique filenames will take some time, too.

psfinaki commented on tests/FSharp.Compiler.ComponentTests/Miscellaneous/MigratedCoreTests.fs:
```diff
@@ -1,465 +1,477 @@
 
-[<Fact>]
-let ``longnames-FSC_OPTIMIZED`` () = singleTestBuildAndRun "core/longnames" FSC_OPTIMIZED
+    [<Fact>]
```
FYI this was turned off quite recently, so it's likely not fixed properly and might start fluking again :/

majocha replied:
yeah. that thing is the horror. I plan to just write a new test, at least for `TryScan`.

psfinaki commented on tests/FSharp.Test.Utilities/Compiler.fs:
```diff
@@ -175,8 +175,13 @@ module rec Compiler =
           StdOut:   string
           StdErr:   string }
 
+    type EvalOutput =
```
I like it. Adding types is the natural F# way to solve problems :)

psfinaki commented on tests/FSharp.Test.Utilities/CompilerAssert.fs:
```diff
@@ -664,7 +656,7 @@ module rec CompilerAssertHelpers =
               member _.Dispose() = try File.Delete runtimeconfigPath with | _ -> () }
 #endif
-        let timeout = 30000
+        let timeout = 60000
```
From my experience, increasing timeouts in tests is rarely effective in the long run. Timeouts usually mean badly designed code, or badly designed tests, or both - so it might be rather just rethink what's going on (not necessarily in this PR)

psfinaki commented:
@majocha great job so far, and impressive results. I added a few notes here and there.

If you happen to fix some other things along the parallelization way, consider rather submitting side-PRs where possible. It's a bit annoying to do that, but that would be helpful for us reviewers and safer from the maintenance perspective.

I'm back from vacation - if you need any help, let me know. At least, I can think about good names for split modules :)

psfinaki commented:
Awesome. Note "Automation.LetBindings.InsideType.Static" is now failing in all PRs, don't mind that one.

majocha commented:
Another obvious problem that I missed: `Environment.CurrentDirectory`
This one is a per process singleton with no workaround at all.
And we have quite a lot of tests accessing it and some scripts writing to it. It all ends up in artifacts/bin folder I guess?

psfinaki commented:
> Another obvious problem that I missed: Environment.CurrentDirectory
This one is a per process singleton with no workaround at all.
And we have quite a lot of tests accessing it and some scripts writing to it. It all ends up in artifacts/bin folder I guess?

Yeah that's where it ends up. I haven't looked at all the cases but I am quite sure that in many of them that's just a quick and often unnecessary hack. We don't really have a consistent story here (surprise), artifacts/bin folder is really a mess:
![image](https://github.com/user-attachments/assets/3c14d3fe-8cd3-4cb3-9831-56f94edbb620)

It's a combination of binaries from "real" projects (hard code on disk, not generated during test execution), and ephemeral ones. Like, we can live with this, but ideally that would sort of match the "real" project structure, whereas everything else would be in-memory, in temp user folders, or here in a separate folder "test stuff". 

But the best of all would be to just clean things up in the tests after execution - again, that would be just applying good coding practices to tests.

Of course, fixing everything at once would be quite a substantial effort - I'm just sharing some guidance on what to do with whatever comes along the way :)

majocha commented:
Yes, it looks like most tests do clean up. It's just that in parallel it means sometimes overstepping on each other. I did a search for `Environment.CurrentDirectory` and it's not as bad as I thought.

I wonder about the script-based tests. They all tend to follow a convention of doing this:

https://github.com/dotnet/fsharp/blob/c228ace3da3f1e90adeac913a30f7b42920d59e5/tests/fsharp/tools/eval/test.fsx#L2800

I guess this is a very old thing. Running in-process that file is probably just constantly recreated in bin?
I wonder, do we still need this?

psfinaki commented:
You're right, it's a very old thing :) 

>  System.IO.File.WriteAllText("test.ok","ok");  

I don't see any good in this line really... That just adds up to IO time and overall burden.

Two more notes on that:
- regardless of how (in the technical sense) those tests report errors, they do it badly: it's always a pain to figure out what's failing there, more often then not I resign on understanding CI logs and try reproing things locally
- our vision is to rewrite all that stuff, hence the guidance is - whenever we need to adjust/fix/add something there, we just rewrite them using a proper testing suite and approach

The point is: it's hard to make things worse there :eyes: we cannot just delete all this (since it's of some value) so we need to keep these tests running _somehow_, but adjusting any underlying machinery is totally fine, and removing useless output (to the console or to files) definitely falls into this category.

psfinaki commented on tests/FSharp.Compiler.ComponentTests/CompilerService/AsyncMemoize.fs:
```diff
@@ -376,7 +375,6 @@ let ``Stress test`` () =
 [<InlineData(true, 1)>]
 [<InlineData(false, 2)>]
 let ``Cancel running jobs with the same key`` cancelDuplicate expectFinished =
-    task {
```
Good stuff, we should have some analyzer for reporting redundant CEs

psfinaki commented on src/Compiler/Utilities/illib.fs:
```diff
@@ -137,7 +137,7 @@ module internal PervasiveAutoOpens =
 
         static member RunImmediate(computation: Async<'T>, ?cancellationToken) =
-            let cancellationToken = defaultArg cancellationToken Async.DefaultCancellationToken
+            let cancellationToken = defaultArg cancellationToken CancellationToken.None
```
What's the reason for this change? It might be a good idea in general, just asking how did you stumble on it.

majocha replied:
This is out of desperation. I'm seeing OperationCancelledException in various test cases. We do a lot of RunSychronously / RunImmediate in the tests but also in FCS. I have no idea how to debug this, because it's very random, but my intuition is that Async.CancelDefaultToken() called in some test cases kills the asyncs in the same process? I have no idea if this makes sense, so this one is just experimental.

psfinaki commented on tests/FSharp.Compiler.ComponentTests/Miscellaneous/FsharpSuiteMigrated.fs:
```diff
@@ -31,16 +31,17 @@ module ScriptRunner =
         match cu with 
         | FS fsSource ->
-            File.Delete("test.ok")
+            // File.Delete("test.ok")
```
Yeah so I've looked into that and I think this is just an old technique to verify that the some command executed without a problem. Modern test framework just explode automatically if this happens.

I might help you out here a bit, removing this at least from tests/fsharp :)

majocha replied:
It'd be great!

psfinaki commented:
Yeah locking is probably the easiest way to go here. Although, depends on how often this happens, it might be worth understanding if that writing is at all necessary - we might be better off not doing that at all.

psfinaki commented:
Some more remarkably useless logs just to keep us motivated :D

![image](https://github.com/user-attachments/assets/22f4a034-446c-4e9f-84b5-fc57739635b7)

majocha commented:
I skipped this test for now:
https://github.com/dotnet/fsharp/blob/3de720beaa378fd297fa3db2b458f212c9b5ef68/tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs#L672-L673

`currentPath` is being shadowed (unintentionally I suppose):
https://github.com/dotnet/fsharp/blob/3de720beaa378fd297fa3db2b458f212c9b5ef68/tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs#L689-L703

 I suspect this test passes by accident only.

I tried to fix it but I'm not sure what the correct behavior should be.

psfinaki commented:
> I skipped this test for now

Damn this just should be two tests, I am sorry we have tests like this 👀

psfinaki commented:
Yeah this is notoriously bad code :) Feel free to switch off the test in a separate PR, maybe someone will understand what's supposed to be happening there. It just compares stuff to an unassigned variable...

majocha commented:
Core.UnitTests now run (mostly) in parallel, too.

There was a global pdb symbols temp folder for fsi session, this is now one per instance. This one probably deserves a seperate PR.

There is erroneous use of `[<CollectionDefinition(..., DisableParallelization  = true)>]` in some places in the tests. This does not work. See https://github.com/xunit/xunit/issues/1999#issuecomment-522635397
I will fix those or remove them as needed.

majocha commented:
Also, I noticed there are already a lot of skipped / flaky `Async` unit tests.
Running in parallel will exacerbate the problem with any tests depending on tight timeouts.

majocha commented:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=807490&view=logs&j=170942dc-5349-5022-2275-77744f335216&t=20776a93-33c2-52f0-ee8f-b8c9809b2e23&l=9001

```
 Failed FSharpChecker.FindReferences.We don't find type aliases for a type [52 ms]
  Error Message:
   System.OperationCanceledException : The operation was canceled.
  Stack Trace:
     at Microsoft.FSharp.Control.AsyncResult`1.Commit() in D:\a\_work\1\s\src\FSharp.Core\async.fs:line 455
   at Microsoft.FSharp.Control.AsyncPrimitives.QueueAsyncAndWaitForResultSynchronously[a](CancellationToken token, FSharpAsync`1 computation, FSharpOption`1 timeout) in D:\a\_work\1\s\src\FSharp.Core\async.fs:line 1143
   at Microsoft.FSharp.Control.FSharpAsync.RunSynchronously[T](FSharpAsync`1 computation, FSharpOption`1 timeout, FSharpOption`1 cancellationToken)
   at FSharpChecker.FindReferences.We don't find type aliases for a type() in D:\a\_work\1\s\tests\FSharp.Compiler.ComponentTests\FSharpChecker\FindReferences.fs:line 347
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

Now I suspect `OperationCanceledException` bubbled up from somewhere in the `TransparentCompiler`.

majocha commented:
I see this one often:
`(Error 193, Line 1, Col 1, Line 1, Col 1, "internal error: Value cannot be null.")`
It will be the hardest to debug, I'm afraid.

psfinaki commented on tests/FSharp.Compiler.ComponentTests/xunit.runner.json:
```diff
@@ -1,7 +1,4 @@
-    "parallelizeTestCollections": false,
-    "maxParallelThreads": 1
+  "$schema": "https://xunit.net/schema/current/xunit.runner.schema.json",
+    "shadowCopy": false
```
Setting this to default true would probably be even more demanding for our tests now, right?

majocha replied:
Yes, but we kind of want to parallelize this one and FSharpSuite at least. This will require some finetuning, for sure.

psfinaki replied:
Sure. Note that _some_ of the component tests are inherently supposed to run sequentially - they read BCL and use memory-mapped files, hence forcing things work in memory will increase the memory consumption of the compiler itself which we don't want.

That was the idea of the suite but then a bunch of things crept in and so this doesn't apply to the whole project anymore, but still might be useful for you to know :)

psfinaki commented on tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj:
```diff
@@ -14,6 +14,13 @@
   </PropertyGroup>
 
+  <ItemGroup>
+    <AssemblyAttribute Include="Xunit.TestFramework">
```
For my understanding, why is this needed and how will this work?

majocha replied:
It installs the customized xUnit TestFramework.
The only customization there is that it installs the multithreaded console capture.
It's a way to guarantee the console is captured before any code in the assembly runs:
https://github.com/dotnet/fsharp/blob/5e7e1e8eec0942853c427dd1dd2c8503fbe7e1fb/tests/FSharp.Test.Utilities/Utilities.fs#L100-L102

psfinaki replied:
Oh I see. Alright, if we decide to stay with this, it's worth documenting somewhere :)

psfinaki commented on tests/FSharp.Test.Utilities/ScriptingShims.fsx:
I wonder if we even need this whole scripting mode anymore... 
I mean, that's unrelated for now, but is probably worth considering dumping for good.

psfinaki commented:
Thanks for the great progress, Jakub :)

On your notes:
> Core.UnitTests now run (mostly) in parallel, too.

🍾 . If you have test runtime comparison at hand, please share :)

> There was a global pdb symbols temp folder for fsi session, this is now one per instance. This one probably deserves a separate PR.

Absolutely, splitting parts of this PR into separate ones would be, as usual, beneficial.

> Also, I noticed there are already a lot of skipped / flaky Async unit tests.

Yeah I reopened the related [issue](https://github.com/dotnet/fsharp/issues/11586), thanks for pointing out. You can probably skip a few more (those in your way), if you don't find a very quick way to fix them.

> Now I suspect OperationCanceledException bubbled up from somewhere in the TransparentCompiler.

Feel free to skip some more TC tests, just point to this [ticket](https://github.com/dotnet/fsharp/issues/16766) there then.

psfinaki commented:
> (Error 193, Line 1, Col 1, Line 1, Col 1, "internal error: Value cannot be null.")

Internal errors are always bad, this just signals we have a problem in the compiler.

And then really pronounces the value of your efforts - redoing things in a proper manner we uncover a lot of shadowed problems we have in the compiler (and looks like even some bugs).

majocha commented:
I'm trying to think how to fine-tune this to not OOM the CI. Other than by trial and error.

psfinaki commented:
OOMs don't reproduce locally, right?

majocha commented:
No, they don't.
However when I run the whole net9.0 target, all FSharp.sln projects in parallel, I can see utilization around 30-40GB and 100%CPU.

psfinaki commented:
I see. Well, it would be nice to at least isolate those which produce OOM in the CI, maybe we can see a common pattern and guess the underlying issue :)

majocha commented:
Yes, unfortunately looks like it's both runs. I have an idea to set xunit.runner to maxthread 1 (basically what we have in main) to just see what happens and then gradually increase the load.

majocha commented:
![image](https://github.com/user-attachments/assets/ebdf7005-f96d-453a-a76d-98fb0b13e0d5)
![image](https://github.com/user-attachments/assets/b3b82890-89dc-4065-9e56-77672ee89706)
Hmm ok. A lucky run, nothing timed out.
This was with  `maxParallelThreads: 2` so there is some room to improvement. Also ComponentTests can still be split into smaller modules to improve CPU utilization more.

psfinaki commented:
Wow! That was just about an hour!

majocha commented:
Yes, I'm not giving up on that.

There's still a lot of room for improvement:
The thing I was not aware is that to run multiple assemblies in parallel you need to launch the whole solution e.g. like this:
`dotnet test .\FSharp.sln  -c Release -f net472 -- xUnit.DiagnosticMessages=true`

In build script it seems we launch test projects in sequence, apart from FSharpSuite that runs in a background task? So the CI run does not get all the benefits yet, but the local runs are much more intense now on the CPU.

psfinaki commented:
That's correct - one of relatively low hanging fruits is to parallelize stuff on the CI level. The CI matrix is already a bit complex and confusing tho, so this should not be making things even more complex. And yeah, it's PowerShell programming 👀

majocha commented:
https://github.com/dotnet/fsharp/blob/f4860a40909b9da3e42f160d7a6205872ced737c/eng/Build.ps1#L603-L614

Now I think this might have something to do with those OOM exceptions in CI that were happening in this PR.
I'd try to not run that as a background task, instead make things straightforward for now.
Run the test projects one by one, but don't limit maxthreads.

majocha commented:
There are two classes of errors that still remain to be dealt with:

1.Timing: timeouts tripping in tests that rely on hardcoded delays like `Async.Sleep 1000`, `.Wait(timeout=...`  but also tests expecting the threadpool threads to be scheduled in one particular order, for example:
https://github.com/dotnet/fsharp/blob/52133d002b14cbf5d16dbd771a7dda23bb29a6d4/tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs#L471-L481

2. Variations on `value cannot be null` that happen during fsi session or checker  initialization. This one shows up under heavy load, when xUnit `maxtreads` are not limited, usually at the beginning of the project test run:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=809681&view=logs&j=170942dc-5349-5022-2275-77744f335216&t=20776a93-33c2-52f0-ee8f-b8c9809b2e23&l=4635



Now what the plan is?
1. can be mitigated by marking the offending tests with `DisableParallelization`. In the long term we can rewrite the offending tests. I was trying to use "orchestration" with synchronization primitives instead of timeouts in some of these tests and it seems to work. Additionally we could use xUnit setting to detect stuck / hanged tests instead of putting timeouts in tests.

2. We do reuse some global checkers in many tests, for example:
https://github.com/dotnet/fsharp/blob/52133d002b14cbf5d16dbd771a7dda23bb29a6d4/tests/FSharp.Test.Utilities/CompilerAssert.fs#L305
 Now what happens when the test methods are started by reflection, in a worker thread, in parallel? Effectively our context is initialized in some static constructors. Or is it? How theradsafe is it all?
I think I'll just put some initialization guard in a xUnit `BeforeAfter` attribute, this cannot be helped.

psfinaki commented:
I think it's a reasonable direction.

> 1. can be mitigated by marking the offending tests with DisableParallelization

Yes, definitely no need to rewrite all the tests properly. Since those tests don't fail all the time but _tend_ to fail, I'd probably just proactively mark all of them to not run in parallel. Bonus points for creating an item and referencing it in the code - because one of things we are doing here is _explaining_ why things are the way the are :)

> 2. We do reuse some global checkers in many tests

Explicit is indeed better than implicit here, reflection is already obscure enough.

majocha commented:
I went for it and modified the `-testDesktop` and `-testCoreClr` in the build script. 
Now these options run the whole FSharp.sln in parallel (respecting the exclusions in _xunit.runner.json_, I hope).

Locally, on a i5 13gen I see runtimes around 4 minutes for `-testCoreClr`, this is a few minutes faster than `main`. `-testDesktop` runs in 18 minutes, I have no idea how it compares to main, I never had the patience :)

psfinaki commented:
Note that since the matters are very subtle here, I'd probably separate CI parallelization also into a separate PR :)

majocha commented:
Yeah, so there are some failures in the tests run by appdomain, but we won't see what happened because xUnit exceptions are not serializable. Oh well.

psfinaki commented:
Which ones do you mean in particular? Maybe I can try hacking something locally.

majocha commented:
In the previous ci run, I think. I'm on the phone right now, so I can't easily find the link, sorry 🙂

psfinaki commented:
Well I guess eventually, all the problems we might be getting here would be due to the fact that we are running in parallel something that shouldn't or cannot work like that yet. So disabling parallelization should heal this as well? Or I am looking at the wrong place?

majocha commented:
@T-Gro, I changed the build script to disable parallelization in ci.
IMV running this conservatively in the CI, and with full speed locally makes the most sense for now.

I'll try to solidify the work here. Concentrating on cleaning this up, reenabling the skipped tests, extracting the ones I modified into separate PRs.

majocha commented:
xUnit gotcha: collection definition must be declared in the same assembly as it is used, not a referenced one.
Also: xUnit documentation comes in the form of answers under Github issues.

psfinaki commented:
> Also: xUnit documentation comes in the form of answers under Github issues.

Yes, indeed, my observation also :)

But - I've been working with xunit and nunit a lot recently and my observation is that xunit is just much more opinionated, that said, my opinions mostly coincide with theirs.

majocha commented:
So that "Internal error: value cannot be null" bug prevents running this in parallel in the CI, as it shows up basically every coreClr run.
However, this is already very usable locally. Running in the IDE or by `dotnet test` things go a bit faster.

psfinaki commented:
We can also disable parallel execution on the specific platform - tho this will diminish the overall benefit of the effort and will require some more script hooks we already have too many of... What are the gains right now, locally? :)

psfinaki commented:
That sounds very hopeful, thanks for sharing the progress :)

majocha commented:
I've named the test runs in the pipeline, might be useful to know at a glance what failed where.
Before:
![image](https://github.com/user-attachments/assets/1e1b9bea-76b6-4c6d-afa0-bc3c5059e089)
After:
![image](https://github.com/user-attachments/assets/bf2a0fb1-a538-433b-bbf4-c7a34f1aaaff)

majocha commented:
Although it seems in the CI the desktop leg is more stable, I can sometimes repro locally
`Warning 193, Line 1, Col 1, Line 1, Col 1, "internal error: Value cannot be null."`
by running `\Build.cmd -c Release -testDesktop`

I've been thinking about why this shows up in the multithreaded run and not in `main`. There still must be some global state in the compiler that is not thread safe. It seems there are lots of global caches in use, implemented as `ConcurrentDictionary`, with non-atomic updates, but does it mean anything? It's not easy to fix without a consistent repro.

psfinaki commented:
One idea that comes to my mind is to isolate a bunch of these tests (that you've seen failing) in one test suite, see unlimited parallelization and then use VS option for running tests until failure. This way the list should narrow down and then we might see something those particular tests have in common.

psfinaki commented on Directory.Build.targets:
```diff
@@ -19,7 +19,7 @@
     <PackageReference Include="xunit" Version="$(XUnitVersion)" />
     <PackageReference Include="xunit.runner.visualstudio" Version="$(XUnitRunnerVersion)" />
-    <PackageReference Include="NunitXml.TestLogger" Version="$(NunitXmlTestLoggerVersion)" />
+    <PackageReference Include="XunitXml.TestLogger" Version="$(XunitXmlTestLoggerVersion)" />
```
Note I am shamelessly stealing this bit for [this PR](https://github.com/dotnet/fsharp/pull/17782/files) :)

majocha replied:
Please, do!

psfinaki replied:
Okay so that one is merged, I resolved conflicts here :)

majocha commented:
Fixed `frameworkTcImportsCache` concurrent mutation.
Looks like the bug is gone now.  I cannot repro locally anymore, CI seems fine, too.

Total tests:
![image](https://github.com/user-attachments/assets/a148ca0a-c7f2-4e61-80dd-6a7314e7bbe8)
CI run time:
![image](https://github.com/user-attachments/assets/0473df0e-bc04-46d2-bd93-20879d38e98e)

psfinaki commented:
This is really, really cool. We'll start getting the sibling PRs in soon and then we'll see if it's worth splitting this further. Thanks for your work :)

majocha commented:
System.TimeoutException : Timeout executing command 'dotnet' 'C:\Users\cloudtest\AppData\Local\Temp\FSharp.Test.Utilities\2024-09-26-958a19b9\dir-0833a591\StackOverflowRepro.exe'


Yep, I've seen this test timeout twice also here: https://github.com/dotnet/fsharp/pull/17795
So it can be related.

OTOH I've tried it on main and it takes like 15 seconds on my PC to run, while the timeout is 30s.

majocha commented:
11 seconds on main:
![image](https://github.com/user-attachments/assets/c352e597-98fb-4a6f-9df6-2c20ad06961d)

11 seconds in this PR:
![image](https://github.com/user-attachments/assets/b63fc9a4-9c58-4ba7-bf59-b3e7b7b98a7a)
At this point I'm just inclined to increase that 30s timeout.

I wonder if it wouldn't be better to get rid of all the timeouts from the tests and use `dotnet test --blame-hang-timeout` instead.

majocha commented:
This is not the end of  it. Super valuable stack trace somehow surfaced:

```
[xUnit.net 00:00:01.02]     FSharp.Compiler.UnitTests.SpanTests.Script_ReadOnlySpanForInDo [FAIL]
[xUnit.net 00:00:01.02]       System.Exception : Error creating evaluation session: System.AggregateException: One or more errors occurred. (Value cannot be null.)
[xUnit.net 00:00:01.02]        ---> System.ArgumentNullException: Value cannot be null.
[xUnit.net 00:00:01.02]          at System.Threading.Monitor.Enter(Object obj)
[xUnit.net 00:00:01.02]          at FSharp.Compiler.AbstractIL.IL.ILPreTypeDefImpl.FSharp.Compiler.AbstractIL.IL.ILPreTypeDef.GetTypeDef() in E:\repos\fsharp\src\Compiler\AbstractIL\il.fs:line 2927
[xUnit.net 00:00:01.02]          at FSharp.Compiler.Import.entities@729-1.Invoke(Tuple`2 tupledArg) in E:\repos\fsharp\src\Compiler\Checking\import.fs:line 731
[xUnit.net 00:00:01.02]          at FSharp.Compiler.Import.multisetDiscriminateAndMap[Key,Value,a](FSharpFunc`2 nodef, FSharpFunc`2 tipf, FSharpList`1 items) in E:\repos\fsharp\src\Compiler\Checking\import.fs:line 689
[xUnit.net 00:00:01.02]          at FSharp.Compiler.Import.ImportILTypeDefList(FSharpFunc`2 amap, Range m, CompilationPath cpath, FSharpList`1 enc, FSharpList`1 items) in E:\repos\fsharp\src\Compiler\Checking\import.fs:line 724
[xUnit.net 00:00:01.02]          at FSharp.Compiler.Import.modty@727.Invoke(Unit _arg2) in E:\repos\fsharp\src\Compiler\Checking\import.fs:line 727
[xUnit.net 00:00:01.02]          at Internal.Utilities.Library.InterruptibleLazy`1.get_Value() in E:\repos\fsharp\src\Compiler\Utilities\illib.fs:line 33
[xUnit.net 00:00:01.02]          at Internal.Utilities.Library.Extras.MaybeLazy`1.Force() in E:\repos\fsharp\src\Compiler\Utilities\lib.fs:line 396
[xUnit.net 00:00:01.02]          at FSharp.Compiler.CompilerImports.TcImports.ccuHasType(CcuThunk ccu, FSharpList`1 nsname, String tname, Boolean publicOnly) in E:\repos\fsharp\src\Compiler\Driver\CompilerImports.fs:line 1296
[xUnit.net 00:00:01.02]          at FSharp.Compiler.CompilerImports.TcImports.tryFindSysTypeCcu@2596-1.Invoke(CcuThunk ccu) in E:\repos\fsharp\src\Compiler\Driver\CompilerImports.fs:line 2596
[xUnit.net 00:00:01.02]          at Microsoft.FSharp.Collections.ArrayModule.loop@1144-40[T](FSharpFunc`2 predicate, T[] array, Int32 i) in E:\repos\fsharp\src\FSharp.Core\array.fs:line 1147
[xUnit.net 00:00:01.02]          at Microsoft.FSharp.Collections.ArrayModule.TryFind[T](FSharpFunc`2 predicate, T[] array) in E:\repos\fsharp\src\FSharp.Core\array.fs:line 1152
[xUnit.net 00:00:01.02]          at FSharp.Compiler.TcGlobals.TcGlobals.tryFindSysTypeCcu(FSharpList`1 path, String nm) in E:\repos\fsharp\src\Compiler\TypedTree\TcGlobals.fs:line 216
[xUnit.net 00:00:01.02]          at FSharp.Compiler.TcGlobals.TcGlobals.tryFindSysAttrib(String nm) in E:\repos\fsharp\src\Compiler\TypedTree\TcGlobals.fs:line 320
[xUnit.net 00:00:01.02]          at FSharp.Compiler.TcGlobals.TcGlobals..ctor(Boolean compilingFSharpCore, ILGlobals ilg, CcuThunk fslibCcu, String directoryToResolveRelativePaths, Boolean mlCompatibility, Boolean isInteractive, Boolean checkNullness, Boolean useReflectionFreeCodeGen, FSharpFunc`2 tryFindSysTypeCcuHelper, Boolean emitDebugInfoInQuotations, Boolean noDebugAttributes, PathMap pathMap, LanguageVersion langVersion, Boolean realsig) in E:\repos\fsharp\src\Compiler\TypedTree\TcGlobals.fs:line 1079
[xUnit.net 00:00:01.02]          at FSharp.Compiler.CompilerImports.TcImports.BuildFrameworkTcImports@2589-10.Invoke(FSharpList`1 _arg10) in E:\repos\fsharp\src\Compiler\Driver\CompilerImports.fs:line 2602
[xUnit.net 00:00:01.02]          at Microsoft.FSharp.Control.AsyncPrimitives.CallThenInvokeNoHijackCheck[a,b](AsyncActivation`1 ctxt, b result1, FSharpFunc`2 userCode) in E:\repos\fsharp\src\FSharp.Core\async.fs:line 528
[xUnit.net 00:00:01.02]          at Microsoft.FSharp.Control.Trampoline.Execute(FSharpFunc`2 firstAction) in E:\repos\fsharp\src\FSharp.Core\async.fs:line 112
[xUnit.net 00:00:01.02]          --- End of inner exception stack trace ---
[xUnit.net 00:00:01.02]          at Internal.Utilities.Library.PervasiveAutoOpens.Async.RunImmediate.Static[T](FSharpAsync`1 computation, FSharpOption`1 cancellationToken) in E:\repos\fsharp\src\Compiler\Utilities\illib.fs:line 151
[xUnit.net 00:00:01.02]          at FSharp.Compiler.Interactive.Shell.FsiEvaluationSession..ctor(FsiEvaluationSessionHostConfig fsi, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, Boolean fsiCollectible, FSharpOption`1 legacyReferenceResolver) in E:\repos\fsharp\src\Compiler\Interactive\fsi.fs:line 4673
[xUnit.net 00:00:01.02]       Stack Trace:
[xUnit.net 00:00:01.02]            at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1448.Invoke(String message)
[xUnit.net 00:00:01.02]         E:\repos\fsharp\src\Compiler\Interactive\fsi.fs(4676,0): at FSharp.Compiler.Interactive.Shell.FsiEvaluationSession..ctor(FsiEvaluationSessionHostConfig fsi, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, Boolean fsiCollectible, FSharpOption`1 legacyReferenceResolver)
[xUnit.net 00:00:01.02]         E:\repos\fsharp\src\Compiler\Interactive\fsi.fs(5007,0): at FSharp.Compiler.Interactive.Shell.FsiEvaluationSession.Create(FsiEvaluationSessionHostConfig fsiConfig, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, FSharpOption`1 collectible, FSharpOption`1 legacyReferenceResolver)
[xUnit.net 00:00:01.02]         E:\repos\fsharp\tests\FSharp.Test.Utilities\CompilerAssert.fs(1004,0): at FSharp.Test.CompilerAssert.RunScriptWithOptionsAndReturnResult(String[] options, String source)
[xUnit.net 00:00:01.02]         E:\repos\fsharp\tests\FSharp.Test.Utilities\CompilerAssert.fs(1019,0): at FSharp.Test.CompilerAssert.RunScriptWithOptions(String[] options, String source, FSharpList`1 expectedErrorMessages)
[xUnit.net 00:00:01.02]         E:\repos\fsharp\tests\FSharp.Test.Utilities\CompilerAssert.fs(1029,0): at FSharp.Test.CompilerAssert.RunScript(String source, FSharpList`1 expectedErrorMessages)
[xUnit.net 00:00:01.02]         E:\repos\fsharp\tests\fsharp\Compiler\Language\SpanTests.fs(91,0): at FSharp.Compiler.UnitTests.SpanTests.Script_ReadOnlySpanForInDo()
[xUnit.net 00:00:01.02]            at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
[xUnit.net 00:00:01.02]            at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
[xUnit.net 00:00:05.32]       failing on main
```

psfinaki commented:
Thanks for adding a nice summary  to the PR. A note on that, currently the number of tests running in the PR is bigger than in main, it would be nice to include some remark on the number change :)

This seems to be getting shape, probably I'd split all those "ok files" manipulations in the separate PR just so that we test it properly and this PR is easier to review.

Note I will be off for two weeks, but the team should be following this.

majocha commented:
Another bummer:
Unhandled `TaskCanceledException` like this:
```fsharp
    [<Fact>]
    let ``unhandled cancel crashes runner`` () =
        async {
            raise (TaskCanceledException())
        }
        |> Async.Start
```
Will simply crash xunit runner.

I suspect `AsyncMemoize` leaks such exceptions from time to time:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=823221&view=logs&j=22fea640-1099-5f32-ec5d-316ad83f359a&t=4607427e-2085-572d-4b25-975ef0413f8f&l=3351

majocha commented:
>MSB4181: The "VSTestTask" task returned false but did not log an error.

https://dev.azure.com/dnceng-public/public/_build/results?buildId=825807&view=logs&j=966a6946-91f0-5397-81f4-d998467880f9&t=344839a9-08c6-5829-65fb-26f32c5a016f&l=6006

I can't think of anything in FSharpSuite that would crash the test run like this. I suspect this is unhandled exception from a background task somewhere in the service.

note: I'm running the tests here still with full parallelism, as a kind of stress-test to weed out issues.

Martin521 commented:
> MSB4181: The "VSTestTask" task returned false but did not log an error.

Under "Tests" you find

```
Error running command 'D:\a_work\1\s\tests\FSharp.Test.Utilities.....dotnet\dotnet.exe' with args 'build /t:RunFSharpScriptAndPrintOutput' in directory 'C:\Users\cloudtest\AppData\Local\Temp\FSharp.Test.Utilities\2024-10-03-2b3e6d18\regression-OverloadResolution-bug-4c505893\regression\OverloadResolution-bug'.
---- stdout below ---
Determining projects to restore...
Restored C:\Users\cloudtest\AppData\Local\Temp\FSharp.Test.Utilities\2024-10-03-2b3e6d18\regression-OverloadResolution-bug-4c505893\regression\OverloadResolution-bug\12c858c5-0ddd-454c-9205-1377e8574746.tmp.fsproj (in 86 ms).

---- stderr below ---
. ERRORLEVEL -1

   at TestFramework.checkResultPassed(CmdResult result) in D:\a\_work\1\s\tests\FSharp.Test.Utilities\TestFramework.fs:line 449
   at TestFramework.execBothToOutCheckPassed@619.Invoke(CmdResult result)
   at TestFramework.execBothToOutCheckPassed@619-1.Invoke(String x)
   at SingleTest.executeFsi@276.Invoke(String testCompilerVersion, String targetFramework) in D:\a\_work\1\s\tests\fsharp\single-test.fs:line 281
   at SingleTest.executeSingleTestBuildAndRun@230.Invoke(OutputType outputType, String compilerType, String targetFramework, Boolean optimize, Boolean buildOnly) in D:\a\_work\1\s\tests\fsharp\single-test.fs:line 282
   at SingleTest.singleTestBuildAndRunCore(TestConfig cfg, String copyFiles, Permutation p, String languageVersion) in D:\a\_work\1\s\tests\fsharp\single-test.fs:line 288
   at SingleTest.singleTestBuildAndRunAux(TestConfig cfg, Permutation p) in D:\a\_work\1\s\tests\fsharp\single-test.fs:line 318
   at SingleTest.singleTestBuildAndRun(String dir, Permutation p) in D:\a\_work\1\s\tests\fsharp\single-test.fs:line 326
   at FSharp.Tests.Core.RegressionTests.OverloadResolution-bug-FSI() in D:\a\_work\1\s\tests\fsharp\tests.fs:line 1640
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

Martin521 commented:
I have no idea though why this works on coreclr and not on desktop

majocha commented:
> Under "Tests" you find for `FSharp.Tests.Core+RegressionTests.OverloadResolution-bug-FSI`

Yes but I suspect this is spurious, an artifact of test host crash. This test passes locally and will pass in a next run. My suspicion is Transparent Compiler produces some unhandled task cancelled exceptions that brings the whole testhost down, as in here https://github.com/dotnet/fsharp/pull/17662#issuecomment-2385549292.

majocha commented:
> I have no idea though why this works on coreclr and not on desktop
What do you mean?

> EDIT: You inserted `| CmdResult.ErrorLevel (msg1, err) -> Assert.Fail (sprintf "%s. ERRORLEVEL %d" msg1 err)` in TestFramework.fs:449
This should be ok as it mimics the code below that was always there.


> EDIT2: This looks suspicious (in Command.exec):
> 
> ```
>                 use reader = File.OpenRead (path)
>                 use ms = new MemoryStream()
>                 do! reader.CopyToAsync (ms) |> (Async.AwaitIAsyncResult >> Async.Ignore)
>                 ms.Position <- 0L
>                 try
>                     do! ms.CopyToAsync(writer.BaseStream) |> (Async.AwaitIAsyncResult >> Async.Ignore)
>                     do! writer.FlushAsync() |> (Async.AwaitIAsyncResult >> Async.Ignore)
>                 with
>                 | :? System.IO.IOException -> //input closed is ok if process is closed
>                     ()
> ```
> 
> Why would one not close the reader and rather swallow all IOExceptions?

I think the reader gets closed by use .. Dispose() but swallowing IO exceptions does not look good indeed.

Martin521 commented:
> Yes but I suspect this is spurious, an artifact of test host crash. This test passes locally and will pass in a next run. My suspicion is Transparent Compiler produces some unhandled task cancelled exceptions that brings the whole testhost down, as in here [#17662 (comment)](https://github.com/dotnet/fsharp/pull/17662#issuecomment-2385549292).

Mmh. The stack trace and error output of this current one seem to point into a different direction. It seems your `Assert.Fail` was hit.

majocha commented:
> Mmh. The stack trace and error output of this current one seem to point into a different direction. It seems your `Assert.Fail` was hit.

Yes, that's a fact it clearly failed and got logged, but not to the console. And still something killed the testhost.

Martin521 commented:
Ah, now I see your point. Thanks.

Martin521 commented:
> > I have no idea though why this works on coreclr and not on desktop
> 
> What do you mean?

Why that specific test succeeded in all test runs (and, it seems, did not hit the Assert.Fail) except the desktop runs.

majocha commented:
> > > I have no idea though why this works on coreclr and not on desktop
> > 
> > 
> > What do you mean?
> 
> Why that specific test succeeded in all test runs (and, it seems, did not hit the Assert.Fail) except the desktop runs. EDIT: sorry for the question, it probably is present only in these runs.

No sorry, all questions are good to gather thoughts.
I think `WindowsCompressedMetadata transparent_compiler_release` where this failed now is actually a `net9.0` only run, despite being named `Windows`.

I'm thinking about how to stabilize these parallel tests in CI. Locally they seem to run very fine to me, it's very hard to repro the issues that show up in CI.

One idea is to deparallelize everything in CI back to how it was. Already I added a lot of exclusions that are not because of inherent shared resources in tests but just because some tests are flaky under load. As in, the shared resource is CPU time.

Another idea is to add a [<FlakyFact>] attribute that would just rerun given test in case of failure, give it another chance or two.
That would allow to remove a lot of parallelism exclusions, resulting in even faster local runs.

It would be still 2-3 times faster locally which is nice.
Another

majocha commented:
Running this with `parallelizeAssembly: true` is a stress-test:
![image](https://github.com/user-attachments/assets/1d6a7dfe-ae02-4a3c-8f86-a4cb25fbc117)

majocha commented:
Green, good merge. The diff is substantially smaller, but I still have to extract all the fixes and modifications of various tests.

Some tests were unskipped to see if they still cause problems and stress-test this, some were fixed and unskipped, some are improved to better behave under higher CPU load. There is also a TryScan Mailbox test I commented out, ugh.

majocha commented:
This started today I think:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=829440&view=logs&j=2f0d093c-1064-5c86-fc5b-b7b1eca8e66a&t=52d0a7a6-39c9-5fa2-86e8-78f84e98a3a2&l=45

```
./build.sh --ci --configuration Release --restore --build --pack --publish -bl /p:SourceBuildNonPortable= /p:ArcadeBuildFromSource=true /p:DotNetBuildSourceOnly=true /p:DotNetBuildRepo=true /p:AssetManifestFileName=SourceBuild_Managed.xml
./build.sh: line 16: /__w/1/s/eng/build.sh: Permission denied
```
No idea what's it about.

majocha commented:
I'm not giving up on this. I'll squash this, clean up a bit and post another draft PR.
I have some wins wrt. utilizing xUnit standard output mechanism and parallel execution of theory cases / collection cases.

T-Gro commented:
> I'm not giving up on this. I'll squash this, clean up a bit and post another draft PR. I have some wins wrt. utilizing xUnit standard output mechanism and parallel execution of theory cases / collection cases.

This is good news @majocha .
If you think this could be enabled per project/collection, this would be a good alternative to postpone solving some of the problems.

(e.g. FSharpSuite is the slowest one at CI, but likely avoids some of the issues because compilation is via separate .exe invocation. Therefore things like shared state inside the compiler should not matter here that much)

majocha commented:
@T-Gro there are a lot of modifications in FSharp.Test.Utilities that I think are in use in basically all test projects so this cannot be just disabled selectively as in: use previous implementation. It can be selectively throttled down, even down to full sequential execution, per project, per module etc. 

I've been timing the test runs locally a bit, what is really problematic and bottlenecked by something is the net472 target. The slowest here for me is ComponentTests, throwing additional cores at it does nothing in net4, there's just no CPU utilization. I suspect it's the thousands of appdomains it creates and unloads.

net9.0 -testCoreClr  runs for me locally in around 4 minutes now:
![image](https://github.com/user-attachments/assets/287d252f-3c35-4924-aa06-a33fa8f1ca78)

What I've been struggling with atm is hanging processes because of files getting locked. For example the test run hangs for 5 minutes and the dump indicates ilread.fs waits to read "System.Security.Cryptography.Primitives.dll" in the dotnet sdk folder. wth?

Anyway, I'll just post what I got in another PR. It'd be good to test this locally on different machines.

T-Gro commented:
Which File IO call was it, was that visible in the stack trace?
We might check if we are using the best set of switches for a read-only operation.

majocha commented:
> Which File IO call was it, was that visible in the stack trace? We might check if we are using the best set of switches for a read-only operation.
It was here IIRC
https://github.com/dotnet/fsharp/blob/20f1408f86736b761fa323a1dff201f3da148627/src/Compiler/AbstractIL/ilread.fs#L250
I don't think I still have the dump file, but I added --blame-hang-timeout to the script so if it reproduces in CI, should be possible to debug.

majocha commented:
Closing in favor of https://github.com/dotnet/fsharp/pull/17872

T-Gro commented:
There might be a race condition in the way the ilModuleReaderCache works, or how the flags are set.
But system dlls should only be ever read from in the build, so it would be good to make it work without shadow copying.
